ftl/common/cache_runner.py

```diff
@@ -0,0 +1,232 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import httplib2
+import json
+import logging
+import requests
+import subprocess
+import sys
+
+from containerregistry.client import docker_name
+from containerregistry.client import docker_creds
+from containerregistry.client.v2_2 import docker_image
+from containerregistry.client.v2_2 import docker_session
+from containerregistry.transport import transport_pool
+
+from ftl.common import ftl_util
+from ftl.common import constants
+
+# we try/catch these imports so we can reuse this code across
+# all cache images, while basing them off individual language
+# builder images
+try:
+    from ftl.php import layer_builder as php_builder
+except ImportError as e:
+    pass
+
+try:
+    from ftl.python import layer_builder as python_builder
+except ImportError as e:
+    pass
+
+PHP = 'php'
+PYTHON = 'python'
+LANGUAGES = [PHP, PYTHON]
+LANGUAGE_CACHES = {
+    PHP: constants.PHP_CACHE_NAMESPACE,
+    PYTHON: constants.PYTHON_CACHE_NAMESPACE
+}
+
+MAPPING_BUCKET = 'ftl-global-cache'
+MAPPING_FILE = '{language}-mapping.json'
+LOCAL_MAPPING_FILE = '/workspace/' + MAPPING_FILE
+
+
+def main():
+    logging.getLogger().setLevel(logging.INFO)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--packages',
+                        action='store',
+                        dest='packages',
+                        nargs='*',
+                        required=True,
+                        type=str,
+                        help='')
+    parser.add_argument('--language',
+                        '-l',
+                        action='store',
+                        dest='language',
+                        required=True,
+                        help='',
+                        choices=LANGUAGES)
+    args = parser.parse_args()
+
+    runner = CacheRunner(args.packages, args.language)
+    runner.populate_cache()
+
+
+class CacheRunner(object):
+    def __init__(self, packages, language):
+        self._packages = packages
+        self._language = language
+
+        _cache = LANGUAGE_CACHES[language]
+        self._cache_name = constants.GLOBAL_CACHE_REGISTRY + '/' + _cache
+
+        self._reg = docker_name.Registry('gcr.io', strict=False)
+        self._creds = docker_creds.DefaultKeychain.Resolve(self._reg)
+        self._transport = transport_pool.Http(httplib2.Http, size=2)
+        self._cache = docker_name.Tag(self._cache_name, strict=False)
+
+        # retrieve mappings when initializing runner
+        self._mappings = self.read_mappings()
+        logging.info(self._mappings)
+
+    def _tag(self, tag):
+        return docker_name.Tag(self._cache_name + ':' + tag)
+
+    def populate_cache(self):
+        existing_entries = self.retrieve_cache_entries()
+        logging.info(existing_entries)
```

Comment:
Could you prefix this logging so it's clear what is being logged? It might make sense to move it into the function, as well.